pkg/client/apiutil/lazyrestmapper/lazyrestmapper.go

```diff
@@ -0,0 +1,252 @@
+		if err = m.addKnownGroupAndReload(input.Group, input.Version); err != nil {
+			return res, err
+		}
+
+		res, err = m.mapper.ResourceFor(input)
+	}
+
+	return res, err
+}
+
+// ResourcesFor implements Mapper.ResourcesFor.
+func (m *LazyRESTMapper) ResourcesFor(input schema.GroupVersionResource) ([]schema.GroupVersionResource, error) {
+	res, err := m.mapper.ResourcesFor(input)
+	if meta.IsNoMatchError(err) {
+		if err = m.addKnownGroupAndReload(input.Group, input.Version); err != nil {
+			return res, err
+		}
+
+		res, err = m.mapper.ResourcesFor(input)
+	}
+
+	return res, err
+}
+
+// RESTMapping implements Mapper.RESTMapping.
+func (m *LazyRESTMapper) RESTMapping(gk schema.GroupKind, versions ...string) (*meta.RESTMapping, error) {
+	res, err := m.mapper.RESTMapping(gk, versions...)
+	if meta.IsNoMatchError(err) {
+		if err = m.addKnownGroupAndReload(gk.Group, versions...); err != nil {
+			return res, err
+		}
+
+		res, err = m.mapper.RESTMapping(gk, versions...)
+	}
+
+	return res, err
+}
+
+// RESTMappings implements Mapper.RESTMappings.
+func (m *LazyRESTMapper) RESTMappings(gk schema.GroupKind, versions ...string) ([]*meta.RESTMapping, error) {
+	res, err := m.mapper.RESTMappings(gk, versions...)
+	if meta.IsNoMatchError(err) {
+		if err = m.addKnownGroupAndReload(gk.Group, versions...); err != nil {
+			return res, err
+		}
+
+		res, err = m.mapper.RESTMappings(gk, versions...)
+	}
+
+	return res, err
+}
+
+// ResourceSingularizer implements Mapper.ResourceSingularizer.
+func (m *LazyRESTMapper) ResourceSingularizer(resource string) (string, error) {
+	return m.mapper.ResourceSingularizer(resource)
+}
+
+// addKnownGroupAndReload reloads the mapper with updated information about missing API group.
+// versions can be specified for partial updates, for instance for v1beta1 version only.
+func (m *LazyRESTMapper) addKnownGroupAndReload(groupName string, versions ...string) error {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	// First, find information about requested group and its versions. Fail immediately if there is no such group.
+	apiGroup, err := m.findAPIGroupByName(groupName)
+	if err != nil {
+		return err
+	}
+
+	// If no specific versions are set by user, we will scan all available ones for the API group.
+	if len(versions) == 0 {
+		for _, version := range apiGroup.Versions {
+			versions = append(versions, version.Version)
+		}
+	}
+
+	// Second, get resources. The number of API calls is equal to the number of versions: /apis/<group>/<version>.
+	groupVersionResources, err := m.fetchGroupVersionResources(apiGroup.Name, versions...)
+	if err != nil {
+		return fmt.Errorf("failed to get API group resources: %w", err)
+	}
+
+	groupResources := &restmapper.APIGroupResources{
+		Group:              apiGroup,
+		VersionedResources: make(map[string][]metav1.APIResource),
+	}
+	for version, resources := range groupVersionResources {
+		groupResources.VersionedResources[version.Version] = resources.APIResources
+	}
```

Comment:
Can we add a fastpath here for how controller-runtime and 99% of everyone who isn't `kubectl` uses this please:

```suggestion
	apiGroup := metav1.APIGroup{
		Name: groupName,
	}
	for _, version := range versions {
		apiGroup.Versions = append(apiGroup.Versions, metav1.GroupVersionForDiscovery{
			GroupVersion: metav1.GroupVersion{Group: groupName, Version: version}.String(),
			Version:      version,
		})
	}
	groupResources := &restmapper.APIGroupResources{
		Group:              apiGroup,
		VersionedResources: make(map[string][]metav1.APIResource),
	}

	if len(versions) == 1 {
		gv := metav1.GroupVersion{
			Group:   groupName,
			Version: versions[0],
		}
		resources, err := m.client.ServerResourcesForGroupVersion(gv.String())
		if err != nil {
			return fmt.Errorf("failed to get Resources for GroupVersion %s: %w", gv.String(), err)
		}
		groupResources.VersionedResources[versions[0]] = resources.APIResources
	} else {
		// First, find information about requested group and its versions. Fail immediately if there is no such group.
		apiGroup, err := m.findAPIGroupByName(groupName)
		if err != nil {
			return err
		}
		groupResources.Group = apiGroup

		// If no specific versions are set by user, we will scan all available ones for the API group.
		if len(versions) == 0 {
			for _, version := range apiGroup.Versions {
				versions = append(versions, version.Version)
			}
		}

		// Second, get resources. The number of API calls is equal to the number of versions: /apis/<group>/<version>.
		groupVersionResources, err := m.fetchGroupVersionResources(apiGroup.Name, versions...)
		if err != nil {
			return fmt.Errorf("failed to get API group resources: %w", err)
		}

		for version, resources := range groupVersionResources {
			groupResources.VersionedResources[version.Version] = resources.APIResources
		}
	}
```

This gets the number of discovery calls down to one per groupVersion.

Reply:
Hey! Thank you for reviewing my code! I was on PTO, so couldn't update it earlier...

I applied the optimization you suggested and it works even better now! I did it slightly differently to keep the code clean and avoid unnecessary conditions. But everything else works as expected - we always do just one API request for each call. Also I added more tests to cover some corner cases.

The changes I introduced in the recent update (besides the tests) affect `addKnownGroupAndReload` only.

Reply:
Awesome, this looks great, thanks so much!

The only thing left at this point from my POV is the point from the other discussion to not have this in its own package but instead construct this in `apiutil.NewDynamicRestMapper` when a new option is passed as Vince suggested here: https://github.com/kubernetes-sigs/controller-runtime/pull/2116#discussion_r1091137091

If you don't want to do that, we can also merge this as-is and someone else will take care of doing that in a follow-up.